docs/USAGE-GUIDE.md

```diff
@@ -500,6 +500,208 @@ legacy SHA-1 algorithms in CertificateVerify messages if TLS1.2 has been negotia
+TLS control messages like session tickets.
+
+#### Sending Application Data
+
+`s2n_send()` and its variants encrypt and send application data to the peer.
+The sending methods return the number of bytes written and may indicate a partial
+write, even if blocking IO is used.
+
+Unlike OpenSSL, repeated calls to `s2n_send()` should not duplicate the original
+parameters, but should update the inputs per the indication of size written.
+
+For example:
+```c
+char data[10]; /* Some data we want to write */
+
+s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+int written = 0;
+do {
+    int w = s2n_send(conn, data + written, sizeof(data) - written, &blocked);
+    if (w < 0 && s2n_error_get_type() != S2N_ERR_T_BLOCKED) {
+        /* Some kind of fatal error */
+        return S2N_FAILURE;
+    }
+    written += w;
+} while (blocked != S2N_NOT_BLOCKED);
+```
+
+`s2n_sendv_with_offset()` behaves like `s2n_send()`, but supports vectorized buffers.
+The offset input should be updated between calls to reflect the data already written.
+
+For example:
+```c
+char data[10]; /* Some data we want to write */
+struct iovec iov[1] = { 0 };
+iov[0].iov_base = data;
+iov[0].iov_len = sieof(data);
+
+s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+int written = 0;
+do {
+    int w = s2n_sendv_with_offset(conn, iov, 1, written, &blocked);
+    if (w < 0 && s2n_error_get_type() != S2N_ERR_T_BLOCKED) {
+        /* Some kind of fatal error */
+        return S2N_FAILURE;
+    }
+    written += w;
+} while (blocked != S2N_NOT_BLOCKED);
+```
+
+`s2n_sendv()` also supports vectorized buffers, but assumes an offset of 0.
+Because of this assumption, a caller would have to make sure that the input vectors
+are updated to account for a partial write. Therefore `s2n_sendv_with_offset()`
+is preferred.
+
+#### Receiving Application Data
+
+`s2n_recv()` reads and decrypts application data from the peer, copying it into
+the application-provided output buffer. It returns the number of bytes read, and
+may indicate a partial read even if blocking IO is used.
+It returns "0" to indicate that the peer has shutdown the connection.
+
+**NOTE:** By default, `s2n_recv()` will return after reading a single TLS record.
+To instead read until the provided output buffer is full, call `s2n_config_set_recv_multi_record()`.
+
+Unlike OpenSSL, repeated calls to `s2n_recv()` should not duplicate the original parameters,
+but should update the inputs per the indication of size read.
+
+For example:
+```c
+char data[10] = { 0 }; /* Empty buffer to read into */
+
+s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+int bytes_read = 0;
+do {
+    int r = s2n_recv(conn, data + bytes_read, sizeof(data) - bytes_read, &blocked);
+    if (r < 0 && s2n_error_get_type() != S2N_ERR_T_BLOCKED) {
+        /* Some kind of fatal error */
+        return S2N_FAILURE;
+    }
+    bytes_read += r;
+} while (blocked != S2N_NOT_BLOCKED);
```

Comment:
It's possible for this loop to return before filling the buffer. For example, the peer might have just written a single byte. It might be worth calling out that the buffer may not be completely filled and the application should read the partial length from `r`.

Reply:
I added print statements to the end of every example using the bytes written/read. Do you think that makes it clear?

Reply:
I think it depends on what the example is trying to accomplish. In this particular case: are we wanting to show how to fill a buffer of 10 bytes, regardless of how it's sent on the wire by the peer? Or are we wanting to show how to read until we get _any_ bytes from the peer.

If it's the former we need to add `&& sizeof(data) != bytes_read`.

If it's the latter, I think it would be clearer to only assign `r` to `bytes_read`, rather than `+=`. Also I would get rid of the `sizeof(data) - bytes_read` and just do `sizeof(data)`.

Reply:
I just added a second example so that I cover both :)

But you can't actually do `&& sizeof(data) != bytes_read` if you want to fill the buffer. I added a warning about that.